api/v1beta1/azureclusteridentity_webhook.go

```diff
@@ -26,3 +30,31 @@ func (c *AzureClusterIdentity) SetupWebhookWithManager(mgr ctrl.Manager) error {
 		For(c).
 		Complete()
 }
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1beta1-azureclusteridentity,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=azureclusteridentities,versions=v1beta1,name=validation.azureclusteridentity.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+
+var _ webhook.Validator = &AzureClusterIdentity{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (c *AzureClusterIdentity) ValidateCreate() error {
+	var allErrs field.ErrorList
+	if c.Spec.Type == UserAssignedMSI && c.Spec.ResourceID == "" {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec", "resourceID"), c.Spec.ResourceID))
+	} else if c.Spec.Type != UserAssignedMSI && c.Spec.ResourceID != "" {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "resourceID"), c.Spec.ResourceID))
+	}
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("AzureClusterIdentity").GroupKind(), c.Name, allErrs)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (c *AzureClusterIdentity) ValidateDelete() error {
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (c *AzureClusterIdentity) ValidateUpdate(oldRaw runtime.Object) error {
+	return nil
```

Comment:
What happens when the user tries to update `AzureClusterIdentity`? 
Is it valid if the user updates `AzureClusterIdentity.Spec.ResourceID` to `""` after creating the resource? I.e. after creating a resource with `AzureClusterIdentity.Spec.ResourceID == UserAssignedMSI` and `AzureClusterIdentity.Spec.ResourceID !== ""`

Reply:
Good question @nawazkh. @willie-yao the validate webhook runs every time an update run after initial create, could be for validating that the field doesn't change when it's immutable but it could also be for enforcing the same validity requirements. In the latter case, we usually share a common validate function between the two, like here https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azurecluster_webhook.go#L135.

I think in this case we don't want the field to be immutable (user might want to rotate identities after cluster is created) but we do want to enforce the same rules that `ResourceID ` is not empty when type is `UserAssignedMSI `.